### PR TITLE
[SPARK-29248][SQL] Add PhysicalWriteInfo with number of partitions

### PR DESCRIPTION
**Pull request by edrevo**
### What changes were proposed in this pull request?
When implementing a ScanBuilder, we require the implementor to provide the schema of the data and the number of partitions.

However, when someone is implementing WriteBuilder we only pass them the schema, but not the number of partitions. This is an asymetrical developer experience. 

This PR adds a `PhysicalWriteInfo` interface that is passed to `createBatchWriterFactory` and `createStreamingWriterFactory` that adds the number of partitions of the data that is going to be written.

### Why are the changes needed?
Passing in the number of partitions on the WriteBuilder would enable data sources to provision their write targets before starting to write. For example:
- it could be used to provision a Kafka topic with a specific number of partitions
- it could be used to scale a microservice prior to sending the data to it
- it could be used to create a DsV2 that sends the data to another spark cluster (currently not possible since the reader wouldn't be able to know the number of partitions)

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Not tested yet

### CODE REVIEW

**Comment by edrevo** [comment]
@rdblue @cloud-fan , I'm opening this PR in response to https://github.com/apache/spark/pull/25945#discussion_r329294622

I haven't been able to implement the change successfully, since I am not familiar with the streaming part of Spark and I am not sure where to get the number of partitions in a streaming execution context. If you can provide me some clues as to where I can find the number of partitions, I'm happy to continue with the code changes.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 1]
```diff
@@ -588,9 +588,12 @@ abstract class StreamExecution(
       table: SupportsWrite,
       options: Map[String, String],
       inputPlan: LogicalPlan): StreamingWrite = {
-    val writeBuilder = table.newWriteBuilder(new CaseInsensitiveStringMap(options.asJava))
-      .withQueryId(id.toString)
-      .withInputDataSchema(inputPlan.schema)
+    val writeInfo = WriteInfoImpl(
+      queryId = id.toString,
+      inputPlan.schema,
+      ???)
```
Comment (edrevo) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: This is where I'm stuck

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala [hunk 2]
```diff
@@ -332,21 +336,37 @@ case class WriteToDataSourceV2Exec(
...
-      .withQueryId(UUID.randomUUID().toString)
+    val writeInfo = WriteInfoImpl(
+      queryId = UUID.randomUUID().toString,
+      query.schema,
+      rdd.getNumPartitions)
+    table.newWriteBuilder(writeOptions, writeInfo)
+  }
+}
+
+trait WriteBase { this: SparkPlan =>
```
Comment (edrevo) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: There's probably a better name for this trait. I'm happy to change it.

File: external/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroTable.scala [hunk 3]
```diff
@@ -42,8 +42,10 @@ case class AvroTable(
   override def inferSchema(files: Seq[FileStatus]): Option[StructType] =
     AvroUtils.inferSchema(sparkSession, options.asScala.toMap, files)
 
-  override def newWriteBuilder(options: CaseInsensitiveStringMap): WriteBuilder =
-    new AvroWriteBuilder(options, paths, formatName, supportsDataType)
+  override def newWriteBuilder(options: CaseInsensitiveStringMap,
```
Comment (cloud-fan) [inline external/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroTable.scala]: nit: the code style should be
```
def fun1(
    arg1: T,
    arg2: T)...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/write/WriteInfoImpl.scala [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.write
+
+import org.apache.spark.sql.types.StructType
+
+private[sql] case class WriteInfoImpl(queryId: String,
+                                      schema: StructType,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/write/WriteInfoImpl.scala]: same as https://github.com/apache/spark/pull/25990/files#r330361054, we should use 4 space intonation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V1FallbackWriters.scala [hunk 5]
```diff
@@ -82,7 +80,7 @@ case class OverwriteByExpressionExecV1(
 }
 
 /** Some helper interfaces that use V2 write semantics through the V1 writer interface. */
-sealed trait V1FallbackWriters extends SupportsV1Write {
+sealed trait V1FallbackWriters extends SupportsV1Write { this: SupportsV1Write =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V1FallbackWriters.scala]: unnecessary change?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/PhysicalWriteInfo.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+import org.apache.spark.annotation.Experimental;
+import org.apache.spark.sql.connector.write.streaming.StreamingDataWriterFactory;
+
+/**
+ * :: Experimental ::
+ * This interface contains physical (i.e. RDD) write information that data sources can use when generating a
+ * {@link DataWriterFactory} or a {@link StreamingDataWriterFactory}.
+ */
+@Experimental
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/PhysicalWriteInfo.java]: we shall use `@Evolving`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/write/WriteInfoImpl.scala [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.write
+
+import org.apache.spark.sql.types.StructType
+
+private[sql] case class WriteInfoImpl(
+    queryId: String,
+    schema: StructType) extends WriteInfo
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/write/WriteInfoImpl.scala]: how about `LogicalWriteInfo`

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/streaming/StreamingWrite.java [hunk 8]
```diff
@@ -48,8 +49,10 @@
    *
    * If this method fails (by throwing an exception), the action will fail and no Spark job will be
    * submitted.
+   *
+   * @param writeInfo Information about the RDD that will be written to this data writer
    */
-  StreamingDataWriterFactory createStreamingWriterFactory();
+  StreamingDataWriterFactory createStreamingWriterFactory(PhysicalWriteInfo writeInfo);
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/streaming/StreamingWrite.java]: for simplicity, we can just call the parameter "info"

**Comment by edrevo** [comment]
@cloud-fan, I have addressed your feedback

**Comment by edrevo** [comment]
Build failure seems unrelated to my changes

**Comment by cloud-fan** [comment]
ok to test

File: external/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroTable.scala [hunk 9]
```diff
@@ -42,8 +42,11 @@ case class AvroTable(
   override def inferSchema(files: Seq[FileStatus]): Option[StructType] =
     AvroUtils.inferSchema(sparkSession, options.asScala.toMap, files)
 
-  override def newWriteBuilder(options: CaseInsensitiveStringMap): WriteBuilder =
-    new AvroWriteBuilder(options, paths, formatName, supportsDataType)
+  override def newWriteBuilder(
+      options: CaseInsensitiveStringMap,
+      writeInfo: LogicalWriteInfo): WriteBuilder =
```
Comment (cloud-fan) [inline external/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroTable.scala]: nit: `info` is good enough as the parameter name.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala [hunk 10]
```diff
@@ -392,18 +392,14 @@ private[kafka010] class KafkaSourceProvider extends DataSourceRegister
     override def newScanBuilder(options: CaseInsensitiveStringMap): ScanBuilder =
       () => new KafkaScan(options)
 
-    override def newWriteBuilder(options: CaseInsensitiveStringMap): WriteBuilder = {
+    override def newWriteBuilder(options: CaseInsensitiveStringMap,
```
Comment (cloud-fan) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala]: nit: 
```
def func(
    para1: T,
    para2: T ...
```

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala [hunk 11]
```diff
@@ -392,18 +392,14 @@ private[kafka010] class KafkaSourceProvider extends DataSourceRegister
...
+                                 writeInfo: LogicalWriteInfo): WriteBuilder = {
```
Comment (cloud-fan) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala]: nit: `info`

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/BatchWrite.java [hunk 12]
```diff
@@ -45,8 +45,10 @@
    *
    * If this method fails (by throwing an exception), the action will fail and no Spark job will be
    * submitted.
+   *
+   * @param info Information about the RDD that will be written to this data writer
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/BatchWrite.java]: nit: `Physical information about the input data that will be written to this data writer.`

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/PhysicalWriteInfo.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+import org.apache.spark.sql.connector.write.streaming.StreamingDataWriterFactory;
+
+/**
+ * This interface contains physical (i.e. RDD) write information that data sources can use when
+ * generating a {@link DataWriterFactory} or a {@link StreamingDataWriterFactory}.
+ */
+@Evolving
+public interface PhysicalWriteInfo {
+  /**
+   * @return  The number of partitions of the RDD that is going to be written.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/PhysicalWriteInfo.java]: `The number of partitions of the input data that is going to be written.` Let's try to not mention RDD, to make the API general.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/write/PhysicalWriteInfoImpl.scala [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.write
+
+case class PhysicalWriteInfoImpl(numPartitions: Int) extends PhysicalWriteInfo
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/write/PhysicalWriteInfoImpl.scala]: `private[sql]`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/StagingInMemoryTableCatalog.scala [hunk 15]
```diff
@@ -88,8 +88,10 @@ class StagingInMemoryTableCatalog extends InMemoryTableCatalog with StagingTable
 
     override def capabilities(): util.Set[TableCapability] = delegateTable.capabilities
 
-    override def newWriteBuilder(options: CaseInsensitiveStringMap): WriteBuilder = {
-      delegateTable.newWriteBuilder(options)
+    override def newWriteBuilder(
+        options: CaseInsensitiveStringMap,
+        writeInfo: LogicalWriteInfo): WriteBuilder = {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/StagingInMemoryTableCatalog.scala]: nit: `info`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V1FallbackWriters.scala [hunk 16]
```diff
@@ -98,22 +96,21 @@ sealed trait V1FallbackWriters extends SupportsV1Write {
...
  * A trait that allows Tables that use V1 Writer interfaces to append data.
  */
-trait SupportsV1Write extends SparkPlan {
-  // TODO: We should be able to work on SparkPlans at this point.
-  def plan: LogicalPlan
-
+trait SupportsV1Write extends SparkPlan with WriteBase {
   protected def writeWithV1(relation: InsertableRelation): RDD[InternalRow] = {
-    relation.insert(Dataset.ofRows(sqlContext.sparkSession, plan), overwrite = false)
+    relation.insert(sqlContext.internalCreateDataFrame(rdd, query.schema), overwrite = false)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V1FallbackWriters.scala]: This is a breaking change. Previously the v1 implementation can get the original query plan but now they can't.

Since we don't need to create writer factory for v1 write, I don't think we need to pass the physical schema to v1 write.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala [hunk 17]
```diff
@@ -84,9 +84,10 @@ case class CreateTableAsSelectExec(
       catalog.createTable(
         ident, schema, partitioning.toArray, properties.asJava) match {
         case table: SupportsWrite =>
-          val writeBuilder = table.newWriteBuilder(writeOptions)
-            .withInputDataSchema(schema)
-            .withQueryId(UUID.randomUUID().toString)
+          val writeInfo = LogicalWriteInfoImpl(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: nit: `info`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala [hunk 18]
```diff
@@ -335,46 +337,53 @@ trait BatchWriteHelper {
...
-      .withInputDataSchema(query.schema)
-      .withQueryId(UUID.randomUUID().toString)
+    val writeInfo = LogicalWriteInfoImpl(
+      queryId = UUID.randomUUID().toString,
+      query.schema)
+    table.newWriteBuilder(writeOptions, writeInfo)
+  }
+}
+
+trait WriteBase { this: SparkPlan =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: If we don't change the v1 write, we don't need to have this trait, right?

**Comment by cloud-fan** [comment]
in general look good. Please make sure `info` is consistently applied as the parameter name.

**Comment by edrevo** [comment]
New round of feedback has been addressed

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/BatchWrite.java [hunk 19]
```diff
@@ -45,8 +45,10 @@
    *
    * If this method fails (by throwing an exception), the action will fail and no Spark job will be
    * submitted.
+   *
+   * @param info Physical information about the input data that will be written to this data writer.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/BatchWrite.java]: `will be written to this data writer.`

I'd say `will be written to this table`. We write one partition to each data writer, but the physical information is not of one partition, but the entire dataset.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * This interface contains write information that data sources can use when generating a
+ * {@link WriteBuilder}.
+ */
+@Evolving
+public interface LogicalWriteInfo {
+  /**
+   * @return `queryId` is a unique string of the query. It's possible that there are many queries
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java]: nit: `@return` is not useful here.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+   * running at the same time, or a query is restarted and resumed. {@link BatchWrite} can use
+   * this id to identify the query.
+   */
+  String queryId();
+
+  /**
+   * @return the schema of the input data from Spark to data source.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java]: ditto

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/PhysicalWriteInfo.java [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+import org.apache.spark.sql.connector.write.streaming.StreamingDataWriterFactory;
+
+/**
+ * This interface contains physical write information that data sources can use when
+ * generating a {@link DataWriterFactory} or a {@link StreamingDataWriterFactory}.
+ */
+@Evolving
+public interface PhysicalWriteInfo {
+  /**
+   * @return  The number of partitions of the input data that is going to be written.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/PhysicalWriteInfo.java]: ditto

**Review by cloud-fan — APPROVED** [review]
LGTM except a few minor comments

**Comment by edrevo** [comment]
Javadoc fixed. Thanks for the review!!

**Comment by cloud-fan** [comment]
LGTM. @rdblue @brkyvz do you have more comments?

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
@edrevo can you update the PR title and description?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * This interface contains write information that data sources can use when generating a
+ * {@link WriteBuilder}.
+ */
+@Evolving
+public interface LogicalWriteInfo {
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/LogicalWriteInfo.java]: Why are write options not part of `LogicalWriteInfo`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 24]
```diff
@@ -27,7 +27,6 @@ import scala.util.{Failure, Try}
 import org.apache.hadoop.conf.Configuration
 import org.apache.hadoop.fs.{FileStatus, Path}
 import org.apache.hadoop.mapreduce._
-import org.apache.hadoop.mapreduce.lib.input.FileSplit
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Nit: this looks like an unrelated change that causes this to touch an extra file. Can you remove it?

**Comment by rdblue** [comment]
What is the benefit of adding `LogicalWriteInfo`? As far as I can tell, this just changes how schema and query ID are passed to the logical write builder. That doesn't seem very useful to me and this would be a much smaller change without it.

I'd prefer to separate the change to use `LogicalWriteInfo` into its own PR and focus this one on getting the number of partitions to the stream writer.

**Comment by cloud-fan** [comment]
This PR starts to add `LogicalWriteInfo` but then finds out that it can't hold number of partitions. Now `LogicalWriteInfo` is not necessary and agree with @rdblue that it's better to have a separated PR.

**Comment by edrevo** [comment]
Continued in https://github.com/apache/spark/pull/26591 Sorry for closing this PR!

**Comment by rdblue** [comment]
@edrevo, thanks for pushing this to completion!
